File: src/387/cpp/4p.cpp
#include <bits/stdc++.h>

using namespace std;

// 题目链接: https://leetcode.cn/problems/distribute-elements-into-two-arrays-ii/description/
class Fenwick {
  vector<int> tree;

 public:
  Fenwick(int n)
	  : tree(n) {

  }

  // 把下标为 i 的元素增加 1
  void add(int i) {
	while (i < tree.size()) {
	  tree[i]++;
	  i += lowbit(i);
	}
  }

  // 返回下标在 [1,i] 的元素之和
  int pre(int i) {
	int res = 0;
	while (i > 0) {
	  res += tree[i];
	  i -= lowbit(i);
	}

	return res;
  }

  int lowbit(int i) {
	return i & (-i);
  }
};

class Solution {
 public:
  vector<int> resultArray(vector<int> &nums) {
	auto sorted = nums;
	ranges::sort(sorted);
	int m = sorted.size();

	vector<int> a{nums[0]}, b{nums[1]};
	Fenwick t1(m + 1), t2(m + 1);
	t1.add(ranges::lower_bound(sorted, nums[0]) - sorted.begin() + 1);
	t2.add(ranges::lower_bound(sorted, nums[1]) - sorted.begin() + 1);

	for (int i = 2; i < nums.size(); ++i) {
	  int x = nums[i];
	  int v = ranges::lower_bound(sorted, x) - sorted.begin() + 1;
	  int gc1 = a.size() - t1.pre(v);
	  int gc2 = b.size() - t2.pre(v);

	  if (gc1 > gc2 || (gc1 == gc2 && a.size() <= b.size())) {
		a.push_back(x);
		t1.add(v);
	  } else {
		b.push_back(x);
		t2.add(v);
	  }
	}

	a.insert(a.end(), b.begin(), b.end());
	return a;
  }
};

auto __FAST_IO__ = []() noexcept {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  return 0;
};

File: src/387/py/4p.py
from bisect import bisect_left
from typing import List


class Fenwick:
    __slots__ = 'tree'

    def __init__(self, n: int):
        self.tree = [0] * n

    # 把下标为 i 的元素增加 1
    def add(self, i: int) -> None:
        while i < len(self.tree):
            self.tree[i] += 1
            i += self.lowbit(i)

    # 返回下标在 [1,i] 的元素之和
    def pre(self, i: int) -> int:
        res = 0
        while i > 0:
            res += self.tree[i]
            i -= self.lowbit(i)

        return res

    def lowbit(self, i: int) -> int:
        return i & (-i)


class Solution:
    def resultArray(self, nums: List[int]) -> List[int]:
        sorted_nums = sorted(nums)
        m = len(sorted_nums)
        a = [nums[0]]
        b = [nums[1]]
        t1 = Fenwick(m + 1)
        t2 = Fenwick(m + 1)
        t1.add(bisect_left(sorted_nums, nums[0]) + 1)
        t2.add(bisect_left(sorted_nums, nums[1]) + 1)

        for x in nums[2:]:
            v = bisect_left(sorted_nums, x) + 1
            gc1 = len(a) - t1.pre(v)
            gc2 = len(b) - t2.pre(v)

            if gc1 > gc2 or gc1 == gc2 and len(a) <= len(b):
                a.append(x)
                t1.add(v)
            else:
                b.append(x)
                t2.add(v)

        return a + b
